<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [4]:
# Импорт библиотек
import pandas as pd
import psycopg2

In [6]:
# Создаем соединение с заданными параметрами
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [7]:
# текст запроса
query_3_1 = f'''select count(id)
                from public.vacancies
             '''

In [8]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\пк\AppData\Local\Temp\ipykernel_8308\1114379196.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Запрос, который посчитает количество работодателей (таблица employers). 

In [9]:
# текст запроса
query_3_2 = f'''select count(id)
                from public.employers
             '''

In [10]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\пк\AppData\Local\Temp\ipykernel_8308\3062646138.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Запрос для подсчета количества регионов (таблица areas).

In [11]:
# текст запроса
query_3_3 = f'''select count(id)
                from public.areas
             '''

In [12]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\пк\AppData\Local\Temp\ipykernel_8308\928149064.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Запрос для подсчета количества сфер деятельности в базе (таблица industries).

In [13]:
# текст запроса
query_3_4 = f'''select count(id)
                from public.industries 
             '''

In [14]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\пк\AppData\Local\Temp\ipykernel_8308\2101691108.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

***

**Предварительный анализ**

*В ходе предварительного анализа были получены следующие результаты:*

- Всего в базе данных содержится 49197 вакансий.
- Количество работодателей достигает 23501.
- В данных информация по 1362 регионам.
- В базе представлены 294 сферы деятельности.

# Юнит 4. Детальный анализ вакансий

1. Запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Сортировка по количеству вакансий в порядке убывания.

In [15]:
# текст запроса
query_4_1 = f'''select
                    a.name, 
                    count(v.id)
                from public.vacancies v
                join public.areas a on a.id = v.area_id
                group by a.name
                order by 2 desc 
                limit 5
             '''


In [16]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\пк\AppData\Local\Temp\ipykernel_8308\2457173557.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


name  count
0           Москва   5333
1  Санкт-Петербург   2851
2            Минск   2112
3      Новосибирск   2006
4           Алматы   1892

2. Запрос, определяющий у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [17]:
# текст запроса
query_4_2 = f'''select 
                    count(v.id)
                from public.vacancies v
                where salary_from is not null or salary_to is not null
             '''

In [18]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\пк\AppData\Local\Temp\ipykernel_8308\970655580.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


count
0  24073

3. Cредние значения для нижней и верхней границы зарплатной вилки, округленное до целого.

In [19]:
# текст запроса
query_4_3 = f'''select 
                    round(avg(salary_from)) "нижняя граница",
                    round(avg(salary_to)) "верхняя граница"
                from public.vacancies v
             '''

In [20]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\пк\AppData\Local\Temp\ipykernel_8308\2500607024.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


нижняя граница  верхняя граница
0         71065.0         110537.0

4. Запрос, который выводит количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортирован по убыванию количества.


In [21]:
# текст запроса
query_4_4 = f'''select 
                    count(v.id) количество,
                    v.schedule "тип графика",
                    v.employment "тип т/устройства"
                from public.vacancies v
                group by v.schedule, v.employment
                order by 1 desc
             '''

In [22]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\пк\AppData\Local\Temp\ipykernel_8308\2022935265.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


количество       тип графика     тип т/устройства
0        35367       Полный день     Полная занятость
1         7802  Удаленная работа     Полная занятость
2         1593     Гибкий график     Полная занятость
3         1312  Удаленная работа  Частичная занятость
4          940    Сменный график     Полная занятость
5          569       Полный день           Стажировка
6          367    Вахтовый метод     Полная занятость
7          347       Полный день  Частичная занятость
8          312     Гибкий график  Частичная занятость
9          141       Полный день     Проектная работа
10         133  Удаленная работа     Проектная работа
11         116     Гибкий график           Стажировка
12         101    Сменный график  Частичная занятость
13          64  Удаленная работа           Стажировка
14          18     Гибкий график     Проектная работа
15          12    Сменный график           Стажировка
16           2    Вахтовый метод     Проектная работа
17           1    Сменный график     Проектная работа

5. Запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [23]:
# текст запроса
query_4_5 = f'''select 
                    v.experience,
                    count(v.id) количество
                from public.vacancies v
                group by v.experience
                order by 2 
             '''

In [24]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\пк\AppData\Local\Temp\ipykernel_8308\2316260449.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience  количество
0         Более 6 лет        1337
1           Нет опыта        7197
2       От 3 до 6 лет       14511
3  От 1 года до 3 лет       26152

***

**Вывод по детальный анализу вакансий**

1. Топ-5 регионов по количеству вакансий:

- Москва
- Санкт-Петербург
- Минск
- Новосибирск
- Алматы

2. Зарплатные вилки:

- Среднее значение нижней границы: 71065.
- Среднее значение верхней границы: 110537.

3. Формат занятости:

Наиболее популярный формат: Полная занятость с полным рабочим днем

4. Требуемый опыт работы:

Самый популярный уровень опыта: От 1 года до 3 лет.


# Юнит 5. Анализ работодателей

1. Запрос, который позволяет узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [25]:
# текст запроса
# выводим запрос для первого места и присоединяем 
# к нему запрос для пятого места
query_5_1 = f'''(select 
                    e.name Работодатель,
                    count(v.id) "Количество вакансий"
                from public.employers e
                join vacancies v on v.employer_id = e.id
                group by e.name
                order by 2 DESC
                limit 1)
    
                UNION ALL  
    
                (select 
                    e.name Работодатель,
                    count(v.id) "Количество вакансий"
                from public.employers e
                join vacancies v on v.employer_id = e.id
                group by e.name
                order by 2 DESC
                offset 4
                limit 1)
             '''

In [26]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\пк\AppData\Local\Temp\ipykernel_8308\3834399874.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


Работодатель  Количество вакансий
0         Яндекс                 1933
1  Газпром нефть                  331

2. Запрос, который для каждого региона выводит количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдем тот, в котором наибольшее количество работодателей.


In [27]:
# текст запроса
query_5_2 = f'''select 
                    a.name регион,
                    count(e.id) "количество работодателей",
                    count(v.id) "количество вакансий"
                from public.areas a
                left join public.employers e ON a.id = e.area
                left join public.vacancies v ON v.area_id = a.id
                where v.area_id is null
                group by 1
                order by 2 DESC, 3  
             '''

In [28]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\пк\AppData\Local\Temp\ipykernel_8308\2250345976.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


регион  количество работодателей  \
0                          Россия                       410   
1                       Казахстан                       207   
2              Московская область                        75   
3              Краснодарский край                        19   
4                        Беларусь                        18   
..                            ...                       ...   
588                      Струнино                         0   
589    Славгород (Алтайский край)                         0   
590                     Сорочинск                         0   
591  Ленинский (Тульская область)                         0   
592                         Тавда                         0   

     количество вакансий  
0                      0  
1                      0  
2                      0  
3                      0  
4                      0  
..                   ...  
588                    0  
589                    0  
590                    0  
591                    0  
592                    0  

[593 rows x 3 columns]

3. Запрос для подсчета количества регионов для каждого работодателя, в которых он публикует свои вакансии. 


In [29]:
# текст запроса
query_5_3 = f'''select 
                    e.name Работадатель,
                    count(distinct v.area_id) "Количество регионов"
                from public.vacancies v
                join public.employers e ON v.employer_id = e.id
                group by 1
                order by 2 DESC
             '''

In [30]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\пк\AppData\Local\Temp\ipykernel_8308\201667002.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


Работадатель  Количество регионов
0                      Яндекс                  181
1                  Ростелеком                  152
2                  Спецремонт                  116
3      Поляков Денис Иванович                   88
4                   ООО ЕФИН                    71
...                       ...                  ...
14761                  UniSol                    1
14762            UNISTORY LLC                    1
14763                   UNIT6                    1
14764     United Distribution                    1
14765              UNITEL LLC                    1

[14766 rows x 2 columns]

4. Запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [31]:
# текст запроса
query_5_4 = f'''select 
                    count(*) Количество
                from public.employers e
                left join public.employers_industries ei on ei.employer_id = e.id
                where ei.industry_id is null
             '''

In [32]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\пк\AppData\Local\Temp\ipykernel_8308\3493105328.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


Количество
0        8419

5. Запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [33]:
# текст запроса
query_5_5 = f'''select
                    e.name Компания,
                    count(e.id) Количество
                from public.employers e
                left join public.employers_industries ei on ei.employer_id = e.id
                group by 1
                having count(e.id) = 4
                order by 1
                offset 2
                limit 1
             '''

In [34]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\пк\AppData\Local\Temp\ipykernel_8308\388711819.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


Компания  Количество
0     2ГИС           4

6. Запрос, который выводит какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [35]:
# текст запроса
query_5_6 = f'''select
                    count(ei.employer_id) Количество
                from public.employers_industries ei
                join public.industries i ON ei.industry_id = i.id
                where i.name = 'Разработка программного обеспечения'
             '''


In [36]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\пк\AppData\Local\Temp\ipykernel_8308\821628714.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


Количество
0        3553

7. Для компании «Яндекс» выведем список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавим строку Total с общим количеством вакансий компании. Результат отсортируем по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 


In [37]:
# код для получения списка городов-милионников 
# импортируем библиотеки
import requests 
from bs4 import BeautifulSoup

# Загружаем страницу
url = 'https://ru.wikipedia.org/wiki/Города-миллионеры_России'
response = requests.get(url, headers={'User-Agent':'Mozilla/5.0'})
soup = BeautifulSoup(response.text, 'html.parser')
table = soup.find('table', class_="standard sortable")
# Создаем список
city_ls = []
if table:
    # Получаем текст с помощью метода find()
    for row in table.find_all('tr'):
        cells = row.find_all('td')
        # Если длина ячейки >= 2
        if len(cells) >= 2:
            city_link = cells[1].find('a')
            if city_link:
                # Добавляем название города в список, удаляя лишние пробелы
                city_ls.append(city_link.get_text(strip=True))
else:
    print("Таблица с городами не найдена")

city_tpl = tuple(city_ls)



In [38]:
# текст запроса
query_5_7 = f'''(select
                    a.name Город,
                    count(v.id) "Количество вакансий"
                from public.vacancies v
                join public.areas a ON v.area_id = a.id
                join public.employers e on v.employer_id = e.id
                where e.name = 'Яндекс'
                and a.name in {city_tpl}
                group by a.name)
                UNION ALL
                (select
                    'total', count(v.id)
                from public.vacancies v
                join public.areas a ON v.area_id = a.id
                join public.employers e on v.employer_id = e.id
                where e.name = 'Яндекс'
                and a.name in {city_tpl})
                order by 2
             '''

In [39]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\пк\AppData\Local\Temp\ipykernel_8308\1965596796.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


Город  Количество вакансий
0              Омск                   21
1         Челябинск                   22
2        Красноярск                   23
3         Волгоград                   24
4             Пермь                   25
5            Казань                   25
6    Ростов-на-Дону                   25
7               Уфа                   26
8            Самара                   26
9         Краснодар                   30
10          Воронеж                   32
11      Новосибирск                   35
12  Нижний Новгород                   36
13     Екатеринбург                   39
14  Санкт-Петербург                   42
15           Москва                   54
16            total                  485

***

**Выводы по анализу работодателей**

1. Яндекс находится на первом месте по количеству вакансий (1933 вакансии)

2. В Москве, Санкт-Петербурге и Минске наибольшее количество работодателей и вакансий. Регион Россия реже всего указывают в вакансиях.

3. Яндекс публикует вакансии в большем количестве регионов 

4. У 8419 работодателей не указана сфера деятельности

5. Компания 2ГИС находится на 3м месте среди компаний у которых указаны 4 сферы деятельности

6. 3553 компаний указали "Разработка программного обеспечения" в сфере своей деятельности

7. Компания Яндекс предствалена в городах миллионниках в общем количестве 485 вакансий 

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [40]:
# текст запроса
query_6_1 = f'''select
                    count(v.id) Количество
                from public.vacancies v
                where v.name ilike '%data%' or 
                      v.name ilike '%данн%'
             '''

In [41]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df


C:\Users\пк\AppData\Local\Temp\ipykernel_8308\495031047.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


Количество
0        1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [42]:
# текст запроса
query_6_2 = f'''select
                    count(v.id) "Количество вакансий"
                from public.vacancies v
                where (v.name ilike '%data scientist%' or 
                      v.name ilike '%data science%' or
                      v.name ilike '%исследователь данных%' or 
                      (v.name like '%ML%' and v.name NOT LIKE '%HTML%') or
                      v.name ilike '%machine learning%' or
                      v.name ilike '%машинн%обучен%')
                      and
                      (v.name ilike '%junior%' OR
                       v.experience ilike '%Нет опыта%' OR
                       v.experience ilike '%Стажировка%')
             '''


In [43]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\пк\AppData\Local\Temp\ipykernel_8308\2267038862.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


Количество вакансий
0                   51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [44]:
# текст запроса
query_6_3 = f'''select
                    count(v.id) "Количество вакансий"
                from public.vacancies v
                where (v.name ilike '%data scientist%' or 
                      v.name ilike '%data science%' or
                      v.name ilike '%исследователь данных%' or 
                      (v.name ilike '%ML%' and v.name not ilike '%HTML%') or
                      v.name ilike '%machine learning%' or
                      v.name ilike '%машинн%обучен%')
                      and
                      (v.key_skills ilike '%SQL%' or v.key_skills ilike '%postgres%')
             '''

In [45]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\пк\AppData\Local\Temp\ipykernel_8308\3794299702.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


Количество вакансий
0                  229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [46]:
# текст запроса
query_6_4 = f'''select
                    count(v.id) "Количество вакансий"
                from public.vacancies v
                where (v.name ilike '%data scientist%' or 
                      v.name ilike '%data science%' or
                      v.name ilike '%исследователь данных%' or 
                      (v.name ilike '%ML%' and v.name not ilike '%HTML%') or
                      v.name ilike '%machine learning%' or
                      v.name ilike '%машинн%обучен%')
                      and
                      v.key_skills ilike '%Python%'
             '''

In [47]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\пк\AppData\Local\Temp\ipykernel_8308\4264665179.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


Количество вакансий
0                  357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [48]:
# текст запроса
query_6_5 = f'''select
                    round(avg(length(v.key_skills) - length(replace(v.key_skills, '\t', ''))+1),2) "Среднее кол-во навыков"
                from public.vacancies v
                where (v.name ilike '%data scientist%' or 
                      v.name ilike '%data science%' or
                      v.name ilike '%исследователь данных%' or 
                      (v.name like '%ML%' and v.name not like '%HTML%') or
                      v.name ilike '%machine learning%' or
                      v.name ilike '%машинн%обучен%')
                      and
                      v.key_skills is not null        
             '''

In [49]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\пк\AppData\Local\Temp\ipykernel_8308\3654210517.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


Среднее кол-во навыков
0                    6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [50]:
# текст запроса
query_6_6 = f'''select
                    round(AVG(COALESCE((v.salary_to + v.salary_from)/2, v.salary_to, v.salary_from))) "Средняя зарплата"
                from public.vacancies v
                where (v.name ilike '%data scientist%' or 
                      v.name ilike '%data science%' or
                      v.name ilike '%исследователь данных%' or 
                      (v.name like '%ML%' and v.name not like '%HTML%') or
                      v.name ilike '%machine learning%' or
                      v.name ilike '%машинн%обучен%')
                      and
                      (v.salary_to is not null OR v.salary_from is not null)
                      and v.experience = 'От 3 до 6 лет'       
             '''


In [51]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\пк\AppData\Local\Temp\ipykernel_8308\519194770.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


Средняя зарплата
0          243115.0

***

**Выводы по предметному анализу**

1. Общее число вакансий, имеющих отношение к данным, состовляет 1771, что говорит высоком спросе на специалистов в сфере анализа данных.

2. Для начинающих (junior) дата-сайентистов доступно 51 подходящая вакансия.

3. Вакансий, где "SQL" или "postgre" указаны как ключевые навыки, составляет 229, что свидетельствует о важности владения этими инструментами.

4. Вакансий, где "Python" указан как ключевой навык, составляет 351, что свидетельствует о его важности как одного из основных инструментов в сфере анализа данных.

5. Среднее количество ключевых навыков, предъявляемых к специалисту, превышает 6, что говорит о необходимости обладания разносторонними знаниями и умениями.

6. Специалистам со стажем от 3 до 6 лет предлагается средняя зарплату в размере 243115, что делает професию финансово привлекательной

# 7. Дополнительные исследования

1. Посмотрим количество вакансий в зависимости от сферы деятельности, где ключевые навыки "Python" и "SQL" и возможностью удаленной работы 

**выведем 10 основных сфер*

In [77]:
# Текст запроса
query_7_1 = '''select
                  i.name "Сфера деятельности",
                  COUNT(v.id) "Количество вакансий"
                from public.industries i
                join public.employers_industries ei ON i.id = ei.industry_id
                join public.vacancies v ON ei.employer_id = v.employer_id
                where v.key_skills like '%Python%' 
                and v.key_skills like '%SQL%' 
                and v.schedule like '%Удаленная работа%' 
                group by i.name 
                order by 2 desc
                limit 10
            '''

In [78]:
# результат запроса
df = pd.read_sql_query(query_7_1, connection)
df

C:\Users\пк\AppData\Local\Temp\ipykernel_8308\2931715827.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_1, connection)


Сфера деятельности  Количество вакансий
0                Разработка программного обеспечения                  209
1  Системная интеграция,  автоматизации технологи...                  136
2  Интернет-компания (поисковики, платежные систе...                  118
3                                               Банк                   47
4                         Коллекторская деятельность                   27
5                               Оптоволоконная связь                   18
6  Маркетинговые, рекламные, BTL, дизайнерские, E...                   17
7                                Фиксированная связь                   16
8                                 Интернет-провайдер                   15
9         Розничная сеть (одежда, обувь, аксессуары)                   15

2. Топ-10 вакансий с возможностью удаленной работы

In [100]:
# текст запроса
query_7_2 = '''select
                  v.name "Название вакансии",  
                  count(*) "количество вакансий"
              from public.vacancies v
              where v.key_skills like '%Python%' 
                and v.key_skills like '%SQL%' 
                and v.schedule like '%Удаленная работа%' 
              group by v.name, v.schedule
              order by 2 desc
              limit 10
            '''

In [101]:
# результат запроса
df = pd.read_sql_query(query_7_2, connection)
df

C:\Users\пк\AppData\Local\Temp\ipykernel_8308\746169716.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_2, connection)


Название вакансии  количество вакансий
0                                      Data Engineer                   18
1                                       Data Analyst                   17
2                               Middle Data Engineer                   11
3                    Site Reliability Engineer (SRE)                    9
4                                 Разработчик Python                    8
5                                 Python разработчик                    8
6                                   Python developer                    8
7              Руководитель направления Data analyst                    7
8  Системный администратор Linux, FreeBSD (Удаленно)                    7
9                                 Программист Python                    6

# Общий вывод по проекту

1. Всего в базе данных содержится 49197 вакансий. Количество работодателей достигает 23501. Указана информация по 1362 регионам в рамках 294 направлений работы
2. Среднее значение нижней границы зарплатной вилки: 71065, верхней границы: 110537. Наиболее популярный формат занятости: Полная занятость с полным рабочим днем. Самый популярный уровень опыта работы: От 1 года до 3 лет.
3. Компания "Яндекс" возглавляет список по количеству вакансий, предлагая 1933 рабочих места. Отдельного внимания заслуживает сфера работы с данными, где насчитывается 1771 вакансия. По данным анализа следует, что навуки Python и SQL являются одними из основных навыков для специалиста. В среднем дата-сайентист должен обладать более чем шестью ключевыми компетенциями.
4. Наибольшее количество вакансий приходится на область разработки программного обеспечения (в т.ч и в рамках удаленной работы)

In [102]:
connection.close()